crypto/mlkem/mlkem.c

```diff
@@ -0,0 +1,1544 @@
+        scalar_mult(out, m++, ar = a);
+        for (j = rank-1; j > 0; --j)
+            scalar_mult_add(out, m++, ++ar);
+    }
+}
+
+static void
+matrix_mult_transpose(scalar *out, const scalar *m, const scalar *a, int rank)
+{
+    const scalar *mc = m, *mr, *ar;
+    int i, j;
+
+    for (i = rank; i-- > 0; ++out)  {
+        scalar_mult(out, mr = mc++, ar = a);
+        for (j = rank; --j > 0; )
+            scalar_mult_add(out, (mr += rank), ++ar);
+    }
+}
+
+/*
+ * Generates a secret vector by using |cbd| with the given seed to generate
+ * scalar elements and incrementing |counter| for each slot of the vector.
+ */
+static __owur
+int gencbd_vector(scalar *out, cbd_t cbd, uint8_t *counter,
+                  const uint8_t seed[MLKEM_RANDOM_BYTES], int rank, mctx *ctx)
+{
+    uint8_t input[MLKEM_RANDOM_BYTES + 1];
+
+    memcpy(input, seed, MLKEM_RANDOM_BYTES);
+    while (rank-- > 0) {
+        input[MLKEM_RANDOM_BYTES] = (*counter)++;
+        if (!cbd(out++, input, ctx))
+            return 0;
+    }
+    return 1;
+}
+
+/*-
+ * Expands the matrix from a seed for key generation and for encaps-CPA.
+ * NOTE: FIPS 203 matrix "A" is the transpose of this matrix, computed
+ * by appending the (i,j) indices to the seed in the opposite order!
+ *
+ * Where FIPS 203 computs $t = A * s + e$, we use the transpose of "m".
+ */
+static __owur
+int matrix_expand(scalar *out, const uint8_t rho[MLKEM_RANDOM_BYTES],
+                  int rank, mctx *ctx)
+{
+    uint8_t input[MLKEM_RANDOM_BYTES + 2];
+    int i, j, ret = 0;
+    EVP_MD_CTX *mdctx = EVP_MD_CTX_new();
+
+    if (mdctx == NULL)
+        goto end;
+    memcpy(input, rho, MLKEM_RANDOM_BYTES);
+    for (i = 0; i < rank; i++) {
+        for (j = 0; j < rank; j++) {
+            input[MLKEM_RANDOM_BYTES] = i;
+            input[MLKEM_RANDOM_BYTES + 1] = j;
+            if (!EVP_DigestInit_ex(mdctx, ctx->shake128_cache, NULL)
+                || !EVP_DigestUpdate(mdctx, input, sizeof(input))
+                || !sample_scalar(out++, mdctx))
+                goto end;
+        }
+    }
+
+    ret = 1;
+end:
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+/*
+ * FIPS 203, Section 5.2, Algorithm 14: K-PKE.Encrypt.
+ *
+ * Encrypts a message with given randomness to the ciphertext in |out|. Without
+ * applying the Fujisaki-Okamoto transform this would not result in a CCA
+ * secure scheme, since lattice schemes are vulnerable to decryption failure
+ * oracles.
+ *
+ * Caller passes storage for |y|, |e1| and |u|.
+ */
+static
+int encrypt_cpa(uint8_t *out, const uint8_t *message, const scalar *m,
+                const scalar *t, const uint8_t *randomness, scalar *y,
+                scalar *e1, scalar *u, vinfo_t vinfo, mctx *ctx)
+{
+    uint8_t counter = 0;
+    uint8_t input[MLKEM_RANDOM_BYTES + 1];
+    scalar e2;
+    scalar v;
+    scalar expanded_message;
+    int rank = vinfo->rank;
+    int compressed_vector_bytes = vinfo->du * rank * DEGREE / 8;
```

Comment:
This is the dropped compressed vector size. It did seem compelling to pre-compute this rather simple expression that's used only in one place, but I did consider it.

If you feel it would be more clear to pre-compute all the related constants and always pass in such data, however small the gain, I am quite open to restoring the constant, and adding a field to the "vinfo" structure. Whatever you/others think best.

Reply:
Since the vinfo structure exists for the purpose of defining constants associated with the variant, I would personally make the change to move this constant there too. But I don't feel so strongly about it that I would insist.